### ISSUE TITLE
Restrict who can push to matching branches option in protected branches prevents automerge

### ISSUE THREAD

**Issue by erickmendonca** [bobvanderlinden/probot-auto-merge#290]
I set `probot-auto-merge` to work on some private repositories and although it worked pretty well on other repositories, it is blocked from merging into repositories with protected branches with `Restrict who can push to matching branches` turned on.

Is this a known issue?

After disabling this option, the bot was able to merge fine.

**Comment by erickmendonca** [comment]
I had some time to investigate the issue and here's the explanation for whoever finds this issue in the future:

GitHub API doesn't allow integration accounts to merge or push into branches if they are both protected and restricted to push. You can add users with permissions to push, but on that feature, you cannot select GitHub Apps or other integration accounts to give them permission to do so. It also doesn't work even if you select the app to have admin access to the repository.

This is a common issue for integrations like `probot-auto-merge` and similar applications because that setup is fairly common on companies. I found similar issues on these integrations dating as of last year and this is still not done on GitHub's side, even though some claim it is on their roadmap.

To temporarily solve this issue on my side (I'm running a fork from `probot-auto-merge`), I had to:

- override the code to use another user account (one that I set up specifically for this purpose)
- add special treatment to when GitHub API responds that a PR is blocked as it might be mergeable and just blocked because the integration is not allowed to merge it).

**TL;DR** this is an issue on GitHub's API itself and you either need to clear the restriction to who can push to this branch or create a workaround to merge as another user.

**Comment by bobvanderlinden** [comment]
> override the code to use another user account (one that I set up specifically for this purpose)

Bleh, that's very unfortunate indeed. This is also not something that Probot recommends. Loosen the permissions sounds like the most viable option for people who want to use probot-auto-merge.

Do you have the fork online somewhere?

**Comment by erickmendonca** [comment]
I have this fork private because it also contains other specific logic, but I can share that specific part if you want.

> Bleh, that's very unfortunate indeed. This is also not something that Probot recommends. Loosen the permissions sounds like the most viable option for people who want to use probot-auto-merge.

I also agree, that's why I didn't push a PR for this. GitHub API itself should enable the integrations to work with these restrictions.

**Comment by erickmendonca** [comment]
On my specific case, I couldn't loose the permissions so I made the workaround for now.

**Comment by rogerluan** [comment]
>you cannot select GitHub Apps or other integration accounts to give them permission to do so

This has been changed on the GitHub side and now that's allowed. See https://github.com/bobvanderlinden/probot-auto-merge/issues/805 for more info :)

I believe this issue can be closed @bobvanderlinden
